api/turing/api/ensemblers_api.go

```diff
@@ -109,18 +111,162 @@ func (c EnsemblersController) UpdateEnsembler(
 			"Ensembler type cannot be changed after creation")
 	}
 
+	pyFuncEnsembler, isPyfunc := ensembler.(*models.PyFuncEnsembler)
+
+	var oldPyFuncEnsembler models.PyFuncEnsembler
+	if isPyfunc {
+		oldPyFuncEnsembler = models.PyFuncEnsembler{
+			RunID:       pyFuncEnsembler.RunID,
+			ArtifactURI: pyFuncEnsembler.ArtifactURI,
+		}
+	}
+
 	if err = ensembler.Patch(request.EnsemblerLike); err != nil {
 		return BadRequest("invalid ensembler configuration", err.Error())
 	}
-
 	ensembler, err = c.EnsemblersService.Save(ensembler)
 	if err != nil {
+		if isPyfunc {
+			_ = c.MlflowService.DeleteRun(context.Background(), pyFuncEnsembler.RunID, pyFuncEnsembler.ArtifactURI, true)
+		}
```

Comment:
yep, that's correct, if the update failed, then it will delete the new mlflow run, the reason i dont handle this error were to avoid confusion, e.g user update, then get failed message that this error happens because we couldn't remove the new run

Reply:
> yep, that's correct, if the update failed, then it will delete the new mlflow run

Ah okay, my bad, I forgot you needed to use `pyFuncEnsembler.RunID` since `ensembler` does not contain a `RunID` field.

> the reason i dont handle this error were to avoid confusion, e.g user update, then get failed message that this error happens because we couldn't remove the new run

Hmm... that's a valid concern but could we instead just wrap that a message saying that the deletion of the new run failed while trying to clean up a failed update?

Reply:
> Hmm... that's a valid concern but could we instead just wrap that a message saying that the deletion of the new run failed while trying to clean up a failed update?

Ahh okay, i'll change the handling, so it'll give a error message saying update failed because cleanup process failed